### PR TITLE
Add some more explicit casts

### LINKED ISSUE #193: bad bitshift in issue1717.p4
The left hand side of a shift has to have a width but this test is doing shifts on integers. I think p4c does some constant folding and the typechecker never sees the bad shift.

### PR DESCRIPTION
**Pull request by hackedy**
Fixes #193, #369, also permits no-op or idempotent casts e.g. (bool)(bool)true.

### CODE REVIEW

**Comment by hackedy** [comment]
Missing some casts for typedef

**Comment by hackedy** [comment]
Missing casts for typedef

**Comment by hackedy** [comment]
However I'm going to remove the newtype<-->newtype casts, which I don't think are spec compliant.

**Review by pataei — APPROVED** [review]

File: lib/checker.ml [hunk 1]
```diff
@@ -2310,7 +2316,7 @@ and cast_ok ?(explicit = false) env original_type new_type =
     let types2 = List.map ~f:snd rec2 in
     casts_ok ~explicit env types1 types2 ||
       type_equality env [] (TypRecord rec1) (TypRecord rec2)
-  | _ -> not explicit && type_equality env [] original_type new_type
+  | _ -> type_equality env [] original_type new_type
```
[the remark this answers is not in the document]
Comment (hackedy) [reply lib/checker.ml:2315]: typedef doesn't exist in the language of types though

**Pull request merged**